src/goto-instrument/contracts/dynamic-frames/dfcc_utils.cpp

```diff
@@ -0,0 +1,577 @@
+    parameter_symbolt new_param_symbol;
+    new_param_symbol.base_name = new_base_name;
+    new_param_symbol.name = new_param_id;
+    new_param_symbol.pretty_name = new_param_id;
+    new_param_symbol.type = old_param.type();
+    new_param_symbol.mode = mode;
+    new_param_symbol.module = module;
+    new_param_symbol.location = location;
+    INVARIANT(
+      goto_model.symbol_table.lookup(new_param_id) == NULL, "name clash");
+    if(goto_model.symbol_table.add(new_param_symbol))
+    {
+      log.error() << "dfcc_utilst::clone_parameters: renamed parameter "
+                  << old_param.get_identifier() << " -> "
+                  << new_param.get_identifier() << " already exists"
+                  << messaget::eom;
+      exit(0);
+    }
+  }
+}
+
+const symbolt &dfcc_utilst::clone_and_rename_function(
+  const irep_idt &function_id,
+  const irep_idt &new_function_id,
+  optionalt<typet> new_return_type = {})
+{
+  std::function<const irep_idt(const irep_idt &)> trans_fun =
+    [&](const irep_idt &old_name) { return new_function_id; };
+
+  std::function<const irep_idt(const irep_idt &)> trans_param =
+    [&](const irep_idt &old_name) { return old_name; };
+
+  std::function<const typet(const typet &)> trans_ret_type =
+    [&](const typet &old_type) {
+      return new_return_type.has_value() ? new_return_type.value() : old_type;
+    };
+
+  std::function<const source_locationt(const source_locationt &)> trans_loc =
+    [&](const source_locationt &old_location) { return old_location; };
+
+  return clone_and_rename_function(
+    function_id, trans_fun, trans_param, trans_ret_type, trans_loc);
+}
+
+void dfcc_utilst::wrap_function(
+  const irep_idt &function_id,
+  const irep_idt &wrapped_function_id)
+{
+  auto &goto_functions = goto_model.goto_functions;
+  auto &symbol_table = goto_model.symbol_table;
+
+  auto old_function = goto_functions.function_map.find(function_id);
+  INVARIANT(
+    old_function != goto_functions.function_map.end(),
+    "dfcc_utilst::wrap_function: function to wrap " + id2string(function_id) +
+      " must exist in the program");
+
+  // Register the wrapped function under the new name
+  std::swap(
+    goto_functions.function_map[wrapped_function_id], old_function->second);
+
+  // Check that new name really exists
+  INVARIANT(
+    goto_functions.function_map.find(wrapped_function_id) !=
+      goto_functions.function_map.end(),
+    "dfcc_utilst::wrap_function: wrapped function " +
+      id2string(wrapped_function_id) +
+      " shall exist in function_map at this point");
+
+  // Remove previous entry
+  goto_functions.function_map.erase(old_function);
+
+  // Check that the slot is empty in function_map
+  INVARIANT(
+    goto_functions.function_map.find(function_id) ==
+      goto_functions.function_map.end(),
+    "dfcc_utilst::wrap_function: the original function " +
+      id2string(function_id) +
+      " shall not exist in function_map anymore at this point");
```

Comment:
Are you sanity-checking the STL?

Reply:
this was ported directly from contracts.cpp L1324-1337 without much thought. But yes it seems like it is checking the STL.

Reply:
fixed